rust/src/ldap/ldap.rs

```diff
@@ -0,0 +1,426 @@
+/* Copyright (C) 2024 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+// written by Giuseppe Longo <giuseppe@glongo.it>
+
+use crate::applayer::{self, *};
+use crate::conf::conf_get;
+use crate::core::{AppProto, Flow, ALPROTO_UNKNOWN, IPPROTO_TCP};
+use ldap_parser::ldap::LdapMessage as LdapParserMessage;
+use ldap_parser::FromBer;
+use nom7 as nom;
+use std;
+use std::collections::VecDeque;
+use std::ffi::CString;
+use std::os::raw::{c_char, c_int, c_void};
+
+use crate::ldap::types::*;
+
+static mut LDAP_MAX_TX: usize = 256;
+
+static mut ALPROTO_LDAP: AppProto = ALPROTO_UNKNOWN;
+
+#[derive(AppLayerEvent)]
+enum LdapEvent {
+    TooManyTransactions,
+}
+
+#[derive(Debug)]
+pub struct LdapTransaction {
+    pub tx_id: u64,
+    pub request: Option<LdapMessage>,
+    pub responses: VecDeque<LdapMessage>,
```

Comment:
the pointer issues you mentioned, are they because of storing a ldap parser message here, while the message was probably doing some pointing into the input data instead of copying? The input data is ephemeral, so unless you make a copy of it first, you can't have pointer (in)to it. The solution would be to create your own structs and copy the fields you need, so they can outlive the lifetime of the input data. Is that the issue you're referring to?

Reply:
Yes, it's that (which is the issue we discussed in person).
In fact, I've created my own LdapMessage structs with the fields needed here:
https://github.com/OISF/suricata/pull/10798/commits/b95c7c35d51229c75def9dbdd95faf45cc7e84b7#diff-68d62050879d651f59a4db94da80cc7e207398df553b216748d7e25b3f6447d5R109

Reply:
Ok, I think that could have been explained better then. It's not "somehow" overwriting C pointers, but instead we're just talking about mismatches in lifetimes of data.

Reply:
Sure, I'll make that description better.